include/zephyr/bluetooth/conn.h

```diff
@@ -991,7 +991,19 @@ struct bt_conn_cb {
  */
 void bt_conn_cb_register(struct bt_conn_cb *cb);
 
-/**
+/** @brief Unregister all connection callbacks.
+ *
+ *  Unregister all callbacks to cleanup connections.
+ */
+void bt_conn_cb_unregister_all(void);
```

Comment:
I have a really tough time with this addition. 

We may register the `bt_conn` callbacks from multiple places in the stack, and this allows an application to deregister them all. Not only does this break basically the entire stack, but since there's no callback for the callback being unregistered, there is no way to handle calls to this function. 

Furthermore, this does not handle callbacks registered with `BT_CONN_CB_DEFINE`.

Reply:
Some previous discussion above: https://github.com/zephyrproject-rtos/zephyr/pull/50046#discussion_r969568993

I agree that the callbacks do not need to be unregistered. We should rather supply functions to unregister callbacks and then the application can choose to unregister their own callbacks if they want to.

Reply:
Why do we need to unregister the callbacks at all? 

`bt_conn_cb_register` can called regardless of the BT device state, and I don't see the state of the callbacks as being related to the `bt_disable` or re-enable at all.

Reply:
It is not needed, and since the stack relies on the callbacks they should not be removed.

If however, the application wants to remove them for whatever reason they should do it themselves.

Reply:
Seems to be reasonable to not unregister these callbacks, will update.

Reply:
> It is not needed, and since the stack relies on the callbacks they should not be removed.
> 
> If however, the application wants to remove them for whatever reason they should do it themselves.

Indeed. We haven't had any requests for that functionality AFAIK though. A workaround if you want to unregister is simply to set the callback, in the callback struct, to `NULL`.

Reply:
Would you say the same for `bt_le_scan_unregister_all()`?

Reply:
Yes. Generally I think it's a bad idea to implicit deregister something that an application, or the stack, has registered. None of them would know when or why these are unregistered, and that will generally just break everything (unless we add callbacks for that, but then we are basically back to just leaving the callbacks as is). 

Callbacks are generally not related to the BT stack state.